network/framework/src/protocols/wire/messaging/v1/metadata.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters,
+    protocols::wire::{
+        handshake::v1::ProtocolId,
+        messaging::v1::{MultiplexMessage, NetworkMessage, RpcResponse},
+    },
+};
+use aptos_config::network_id::NetworkId;
+use serde::{Deserialize, Serialize};
+use std::time::SystemTime;
+
+/// A simple struct that wraps a network message with metadata.
+/// Note: this is not sent along the wire, it is only used internally.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct NetworkMessageWithMetadata {
+    /// The metadata about the message
+    message_metadata: MessageMetadata,
+
+    /// The network message to send along the wire
+    network_message: NetworkMessage,
+}
+
+impl NetworkMessageWithMetadata {
+    pub fn new(message_metadata: MessageMetadata, network_message: NetworkMessage) -> Self {
+        Self {
+            message_metadata,
+            network_message,
+        }
+    }
+
+    /// Converts the message into a multiplex message with metadata
+    pub fn into_multiplex_message(self) -> MultiplexMessageWithMetadata {
+        MultiplexMessageWithMetadata::new(
+            self.message_metadata,
+            MultiplexMessage::Message(self.network_message),
+        )
+    }
+
+    /// Consumes the message and returns the individual parts
+    pub fn into_parts(self) -> (MessageMetadata, NetworkMessage) {
+        (self.message_metadata, self.network_message)
+    }
+
+    /// Returns a reference to the message metadata
+    pub fn network_message(&self) -> &NetworkMessage {
+        &self.network_message
+    }
+}
+
+/// A simple struct that wraps a multiplex message with metadata.
+/// Note: this is not sent along the wire, it is only used internally.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct MultiplexMessageWithMetadata {
```

Comment:
do we need Serialize/Deserialize for the structs that's not sent over the wire? it can be more clear to the readers if it doesn't derive them

Reply:
Aah, this was just copy-pasta 😄 Removed serde from everything in this file.